### PR TITLE
Maintenance updates

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context
Update app according to [DataEng Application Maintenance](https://mitlibraries.atlassian.net/wiki/spaces/IN/pages/3445948428/DataEng+Application+Maintenance). 

Main updates were to: 
* Upgrade to Python 3.12
* Update dependencies
* Address new linting requirements

It's worth noting that updates to `mypy` (now version 1.13.0) raised the "No return statement" error for the [`OGMRepository.get_modified_records` function](https://github.com/MITLibraries/geo-harvester/pull/316/files#diff-691ee6bab1bcc33b75cbe92fdb4c17cdf5e92c4de5e34779d729ac413bffec17L247-L268). The linter was confused by the function returning an empty list (the `return []` statements when the first two if conditions are met). 
* **Solution**: Remove the list (`[]`) from the return statement and simply add `return`. Without understanding the nitty-gritty of why this works, the important takeaway is that the return statements terminates the generator (the equivalent of raising a `StopIteration`). When a generator terminates before it can yield any values, if passed to a `list` call, it will return an empty list.

### How can a reviewer manually see the effects of these changes?

1. Run `make lint` and `make test` and verify all linters and unit tests are passing.
2. Review the following **successful** StepFunction executions in Dev1:
   * [`gismit` "full" harvest](https://us-east-1.console.aws.amazon.com/states/home?region=us-east-1#/v2/executions/details/arn:aws:states:us-east-1:222053980223:execution:timdex-ingest-dev:41cd3846-8f28-4e5a-9473-575fd7582dd0) (See[ CloudWatch log](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/timdex-geoharvester-dev/log-events/geo-harvester$252Ftimdex-geoharvester$252Fbfd41a3e6b7c4a8298eb4aa0219a5244), took about 1 hr 10 mins to run)
   * [`gisogm` "daily" harvest](https://us-east-1.console.aws.amazon.com/states/home?region=us-east-1#/v2/executions/details/arn:aws:states:us-east-1:222053980223:execution:timdex-ingest-dev:37971e34-8ce5-497d-814c-f2d01f3fda4c) (See [CloudWatch log](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/timdex-geoharvester-dev/log-events/geo-harvester$252Ftimdex-geoharvester$252F463ea070d234489bb869972e0ca6797c))

### Includes new or updated dependencies?
YES

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- [Include links to Jira Software and/or Jira Service Management tickets here.
](https://mitlibraries.atlassian.net/browse/IN-1088)

### Developer
- [x] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [x] All related Jira tickets are linked in commit message(s)
- [x] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed and verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

**Review by ehanson8 — APPROVED** [review]

File: harvester/harvest/ogm.py [hunk 1]
```diff
@@ -328,7 +329,7 @@ def _get_commit_before_date(self, target_date: str) -> pygit2.Commit | None:
         from_timestamp = date_parser(target_date).timestamp()
         commits = [
             commit
-            for commit in local_repo.walk(local_repo.head.target, pygit2.GIT_SORT_TIME)
+            for commit in local_repo.walk(local_repo.head.target, SortMode.TIME)
```
Comment (ghukill) [inline harvester/harvest/ogm.py:332]: Interesting!  How did this present itself?  In the new version of `pygit2`, was `.GIT_SORT_TIME` not present anymore?

File: harvester/records/formats/marc.py [hunk 2]
```diff
@@ -19,7 +19,7 @@
 # Coordinate string expected in format "hdddmmss" (hemisphere-degrees-minutes-seconds)
 #   - e.g. W1800000, E1800000, N0840000, N0840000
 #   - https://www.loc.gov/marc/bibliographic/bd034.html
-COORDINATE_STRING: TypeAlias = str
+type COORDINATE_STRING = str
```
Comment (ghukill) [inline harvester/records/formats/marc.py:22]: Also interesting, good to know about this syntax.

**Review by ghukill — APPROVED** [review]
Looks good to me!  I was able to run tests and linting locally.

File: harvester/harvest/ogm.py (hunk 1 above)
Comment (jonavellecuerdo) [reply harvester/harvest/ogm.py:332]: I think `mypy` raised the error saying the `.GIT_SORT_TIME` was not available. 🤔

**Pull request merged**